plugins/inputs/logfile/tail/tail_queue.go

```diff
@@ -0,0 +1,43 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT
+
+package tail
+
+import (
+	"github.com/aws/amazon-cloudwatch-agent/plugins/inputs/logfile/fdlimit"
+)
+
+/*
+	Centralize all the tailers which are scheduled for each file
+*/
+
+var defaultOpenFileLimit int = 1024
+
+type TailerEnqueue struct {
+	Queue chan bool
+}
+
+func NewTailerFifoQueue() *TailerEnqueue {
+	if allowedOpenFileLimit, err := fdlimit.CurrentOpenFileLimit(); err == nil {
+		defaultOpenFileLimit = allowedOpenFileLimit
+	}
+
+	return &TailerEnqueue{Queue: make(chan bool, defaultOpenFileLimit)}
+}
+
+func (t *TailerEnqueue) Enqueue() {
+	t.Queue <- true
+}
+
+func (t *TailerEnqueue) Dequeue() bool {
+	select {
+	case v := <-t.Queue:
+		return v
+	default:
+		return false
+	}
```

Comment:
Does this account for the scenario where the channel is empty we try to dequeue?

Reply:
Yes. Default `return false` always ensure even if the channel does not receive any message, the process still return false (since we don't use the value as you mentioned before) and will be garbage collected in some ways. Its the same as
```
if len(enqueue): return false
return true
```